Button to export Cubeviz movie

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request is to add functionality to export Cubeviz slices into a movie file.

Doc: https://jdaviz--2264.org.readthedocs.build/en/2264/cubeviz/plugins.html#movie

![Screenshot 2023-06-29 165809](https://github.com/spacetelescope/jdaviz/assets/2090236/6fbc7ea4-1212-4512-8191-29d2ada55d65)

Example video generated from flux viewer: https://www.youtube.com/watch?v=KticlwspGV4

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #2245 and fixes https://github.com/spacetelescope/jdaviz/issues/2271

### TODO

- [x] Expose FPS to user.
- [x] Baby Shark roundtrip in concept notebook.

### After merge

- Open follow-up issue about GIF support.
- Open follow-up issue about https://github.com/spacetelescope/jdaviz/pull/2264#issuecomment-1612052525

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +85,88 @@ def vue_save_figure(self, filetype):
+
+                # FIXME: bqplot stuck after first frame!
+                # If we can fix this, maybe we can have callback to video.write and don't need PNG files.
+                self.save_figure(filename=cur_pngfile, filetype="png")
```
@mariobuikhuizen or @maartenbreddels , I don't understand why `figure.get_png_data()` gets stuck after the first frame is written out. It never seems to get into `finally`. I tried to brute force setting `figure._upload_png_callback = None` but that did not help at all. What is going on here? 🙏 

```python
from jdaviz import Cubeviz

cubeviz = Cubeviz()
cubeviz.load_data(...)
cubeviz.show()

plg = cubeviz.plugins["Export Plot"]
plg._obj.save_movie(1000, 1010, filename="zzzz.mp4", rm_temp_files=False)
```

```
----> 1 plg._obj.save_movie(1000, 1010, filename="zzzz.mp4", rm_temp_files=False)

File jdaviz/configs/default/plugins/export_plot/export_plot.py:145, in ExportViewer.save_movie(self, i_start, i_end, filename, filetype, rm_temp_files)
    143 slice_plg._on_slider_updated({'new': i})
    144 cur_pngfile = f"._cubeviz_movie_frame_{i}.png"
--> 145 self.save_figure(filename=cur_pngfile, filetype="png")
    146 temp_png_files.append(cur_pngfile)
    147 i += i_step

File jdaviz/configs/default/plugins/export_plot/export_plot.py:76, in ExportViewer.save_figure(self, filename, filetype)
     74             with open(os.path.expanduser(filename), 'bw') as f:
     75                 f.write(data)
---> 76         viewer.figure.get_png_data(on_img_received)
     77 elif filetype == "svg":
     78     viewer.figure.save_svg(filename)

File bqplot/figure.py:207, in Figure.get_png_data(self, callback, scale)
    195 '''
    196 Gets the Figure as a PNG memory view
    197 
   (...)
    204     Scale up the png resolution when scale > 1, when not given base this on the screen pixel ratio.
    205 '''
    206 if self._upload_png_callback:
--> 207     raise Exception('get_png_data already in progress')
    208 self._upload_png_callback = callback
    209 self.send({'type': 'upload_png', 'scale': scale})

Exception: get_png_data already in progress
```

pllim replied:
Thanks!

pllim commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +104,138 @@ def vue_save_figure(self, filetype):
+                # default to MPEG-4
+                filetype = "mp4"
+
+        if filetype != "mp4":
+            raise NotImplementedError(f"filetype={filetype} not supported")
```
To also support GIF, yet another dependency (`imageio`) needs to be introduced. I'd rather not have to tackle that in this PR. I think it would be cleaner as a follow-up PR if we really want that.

pllim commented:
devdeps failure is unrelated; I reported it upstream at https://github.com/newville/asteval/issues/120

camipacifici commented:
Looks very nice, thank you!
A few thoughts:
- There should be a note at the top of the "Export video" to mention the needed package and a link to its installation instructions.
- it would be useful to be able to select the viewer (I created a smoothed cube and opened it in the uncertainty viewer and to save that one out I had to load it in the flux viewer first. I would be easier to just select the viewer I want)
- In standalone, I could not find the movie. Maybe a note to tell the user that they have to set the full path (rev: I actually tried, I misspelled the path, so the path did not exist, but it said it created the movie anyway at /private/var/folders/js/kn36084x6k988161dxy2fwlw0001hl/T/tmp4icsvifz/my-misspelled-path. If it created something, I have no idea where it is ahah)? Or a file picker would be even better.
- I would like to have the option to set the delay between each frame. If I play the movie on my machine, I can play it at a different speed, but if I send this to a journal, I do not think they can set the speed arbitrarily.
Thank you!

camipacifici commented:
Ah! Revision: it followed the viewer I set for export plot! It is ok then, maybe we just need Jenn to make the UI more intuitive.

camipacifici commented:
User says it should also show the axis, please.

pllim commented:
> option to set the delay between each frame

I can present it as [frame rate](https://en.wikipedia.org/wiki/Frame_rate) in FPS or absolute delay in milliseconds between frames. I think the former is more common in video software and gaming, but either is possible. Please pick one.

> User says it should also show the axis, please.

Not sure what this means. It already does. Please see https://www.youtube.com/watch?v=KticlwspGV4 (also linked in original post above).

I'll see if I can address the rest in a follow up commit. Thanks for the quick response!

pllim commented:
> User says it should also show the axis, please.

This turns out to be a bug in voila + bqplot's save_png. You can see the axis missing even when you simply save one PNG file in standalone app, so the fix for this is out of scope.

pllim commented:
> User says it should also show the axis, please.

I don't know why but turns out when in standalone app, the axes are written out as black text on black background, rendering them invisible. If you use the "save PNG" feature to snap one frame, you can see that the axes text is black and transparent background. Is it following my OS theme (which is dark) instead of voila theme (light mode when this happened). If I start the standalone app using `--theme=dark`, I can see the axes: https://www.youtube.com/watch?v=_qwVtpycGUo

pllim commented:
> In standalone, I could not find the movie

This is the same problem in other plugins, see https://github.com/spacetelescope/jdaviz/issues/2271, so I think fixing this is out of scope here, but I can add a note about having to provide full path in standalone app.

camipacifici commented:
> to provide full path in standalone app

it should give an error if the path does not exist.

pllim commented:
> it should give an error if the path does not exist.

Errr... Always error? Or just in standalone app? If the latter, is there a way to detect if we are in standalone app or not?

pllim commented on docs/cubeviz/plugins.rst:
```diff
@@ -274,3 +274,22 @@ Export Plot
+
+.. note::
+
+    For MPEG-4, this feature needs ``opencv-python`` to be installed;
```
@camipacifici , is a note here sufficient?

pllim replied:
So, you want all errors go to this traitlet thingy instead of snackbar?

camipacifici replied:
I'd like to see the note in the plugin itself at a minimum, together with the link to the docs.
I did see the red snackbar when I first used it without having installed the additional package.

pllim replied:
First thing first... do you not see this when opencv-python is not installed and you click "export to MP4"?

![Screenshot 2023-06-29 120827](https://github.com/spacetelescope/jdaviz/assets/2090236/cef6fb75-ee35-440e-a347-5f70fca02231)

pllim replied:
How about this? Note that I cannot insert HTML links into the text, so I can just do plain text.

![Screenshot 2023-06-29 124030](https://github.com/spacetelescope/jdaviz/assets/2090236/d39f068e-6136-4b0c-a578-f0e800faa7fa)

pllim commented:
I think I have addressed the comments as my ability allows. I have noted follow-up issues to open for the things I cannot fix here. See "todo" in original post. Thanks!

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +105,153 @@ def vue_save_figure(self, filetype):
+        if i_end > slice_plg.max_value:
+            i_end = slice_plg.max_value
+        if i_end <= i_start:
+            raise ValueError(f"No frames to write: i_start={i_start}, i_end={i_end}")
```
Any case where an exception can be raised needs to return a functional app (not sure if there are others or not... perhaps the giant try except in `vue_export_movie`?)

```suggestion
            self.app.loading = False
            raise ValueError(f"No frames to write: i_start={i_start}, i_end={i_end}")

```

pllim replied:
Oh, oops. Good catch. I'll accept the suggestion first and see if a simple clean-up is possible later.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +105,153 @@ def vue_save_figure(self, filetype):
+            self.save_figure(filename=cur_pngfile, filetype="png")
+            temp_png_files.append(cur_pngfile)
+            i += i_step
+            time.sleep(ts)  # Avoid giving user epilepsy
```
pulling this from the slice plugin feels weird to me and could result in movies taking forever to export with no clear reason to the user.  Is no sleep really too fast?  Is epilepsy really a concern for realistic data, a non-fullscreen-viewer, and a greyed out application?

rosteen replied:
This answered the first question I had when doing my first test run - I'm trying to save out the whole movie, and it's taking forever (ok, not forever, but 5-10 minutes) to step through the cube, and I wasn't sure if that was really the time it's taking to write the movie or not. I guess it's due to this sleep and maybe writing the movie is much faster? Maybe it's ok to have a sleep here, but I think it could be 5-10x faster than it is now.

kecnry replied:
My preference would be to remove the sleep for now and re-introduce it if it ever becomes a problem (although I'd prefer to just further gray-out and save the time).  The other option might be to pull the sleep from the FPS setting in this plugin rather than the slice plugin?

pllim replied:
Okay I can disable that. But the second sleep to wait for bqplot has to stay.

kecnry replied:
Yes, but maybe something hardcoded and very small?

pllim replied:
Mario originally proposed 0.1 sec per loop, is that small enough?

rosteen replied:
I deleted this sleep and changed the next one to 0.05 and it seemed more reasonable. Definitely not too fast.

kecnry replied:
Honestly even just sitting in a while-pass loop without a sleep would probably be fine, but anything not-noticeable (sub 0.1 sec) should work too.

pllim replied:
Extra sleep is removed. The remaining sleep is now 0.05 as Ricky suggested.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.vue:
```diff
@@ -30,6 +30,67 @@
+                  ></v-text-field>
+                </v-col>
+              </v-row>
+              <v-row>
```
```suggestion
              <v-row justify='end'>

```

kecnry reviewed (commented):
Seems to work well once I read the docs and installed the dependency, although I do think we're doing overkill in artificially extending the time it takes to export.

Do we want a way to cancel if it is taking forever and the user wants to return to the app instead of waiting for it to complete?

I also added some general styling suggestions.  The inputs would also benefit from validation (to prevent negative values for slice numbers (unless that is supported?) and frame rate, require max > min, etc.

rosteen commented on docs/cubeviz/plugins.rst:
```diff
@@ -274,3 +274,23 @@ Export Plot
+directory. Any existing file with the same name will be silently replaced.
+
+When you are ready, click the :guilabel:`Export to MP4` button.
+The movie will be recorded at 5 frames per second. While recording is in progress,
```
Is the 5 FPS just a default or is it hard-coded? In the UI it looks like you can change it (vs the implication here that you can't change it), but I didn't try changing before writing out my first test movie.

pllim replied:
Ooops, doc is outdated. Good catch!

pllim replied:
Fixed.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +105,153 @@ def vue_save_figure(self, filetype):
+            for cur_pngfile in temp_png_files:
+                os.remove(cur_pngfile)
+
+    def save_movie(self, i_start, i_end, fps=5, filename=None, filetype=None, rm_temp_files=True):
```
That same problem exists now in the UI then, which is why I think we need a way to be able to interrupt and kill the thread (and/or don't block the entire app when building the movie).

pllim replied:
The API now grabs from plugins inputs if none provided. And there is a kill switch now.

camipacifici commented:
Tested in standalone again. If I input just a file name, it saves as expected in the directory where I started the tool. If I propose a path (like "../filename") it goes to /private/var/folders/js/kn36084x6k988161dxy2fwlw0001hl/T/, although the path is technically valid.
If I set an invalid path, it correctly sends an error message.

pllim commented:
> If I propose a path (like "../filename")

I am not sure how to solve this for the standalone app in a sane way. If I handle ".." what about ".././../.."? This is a slippery slope.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -20,16 +33,41 @@ class ExportViewer(PluginTemplateMixin, ViewerSelectMixin):
     @property
     def user_api(self):
+        if self.config == "cubeviz":
+            return PluginUserApi(self, expose=('viewer', 'save_figure', 'save_movie'))
```
```suggestion
            return PluginUserApi(self, expose=('viewer', 'save_figure', 'save_movie',
                                                                         'i_start', 'i_end', 'movie_fps', 'movie_filename'))

```

(API docstring would also need to be updated to include these)

pllim replied:
Why? What if someone uses the API to set an invalid value of `i_start` and `i_end` combo? Just more things we have to guard against if they are public. Is `save_movie` not enough? It takes `i_start` and `i_end` as arguments.

kecnry replied:
Because we are trying to strive for reproducible plugin workflows through the API and all other plugins do so by exposing all input widgets.  I get that it is possible here to do the same thing by passing the kwargs, but now the user can't even pull the values set in the widgets.

pllim replied:
Is there a way to have them pull it as read-only? Or you want them to be settable via API on purpose?

kecnry replied:
The rest of the plugin APIs allow setting all public traitlets and there is no current way to expose them as read-only, although I suppose it would be possible to create an infrastructure for that.  Since the method checks the values anyways, what would be the harm in letting the user set them via traitlets?

pllim replied:
I added this as part of a larger change but I'll add you as co-author. I think it is ok to let user change it, because I also added validation to Vue.js so hopefully enough warning that way.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -20,16 +33,41 @@ class ExportViewer(PluginTemplateMixin, ViewerSelectMixin):
+                # NOTE: HTML tags do not work here.
+                self.movie_msg = 'Please install opencv-python to use this feature.'
+
+    def _on_cubeviz_data_added(self, msg):
+        # NOTE: This needs revising if we allow loading more than one cube.
+        if isinstance(msg.viewer, BqplotImageView):
+            if len(msg.data.shape) == 3:
+                self.i_end = msg.data.shape[-1] - 1  # Same as max_value in Slice plugin
```
I wonder if we could have the slice plugin emit a message or have this watch for `slice_plugin.max_value` instead of having to reproduce the logic?

pllim replied:
This is a one-time set up when data is loaded. To listen to a message the entire time app is running for this one call seems excessive, especially since we only allow one cube to be loaded anyway.

kecnry replied:
This currently listens to any `AddDataMessage` and then just filters to `BqplotImageView` (which will then also be triggered any time that the user loads a moment map or other output into either the flux or uncertainty viewer, but might just then be ignored by the shape check).  My concern is just that this logic is duplicated and we could easily introduce a bug by updating it in one place but not the other.  If you don't think listening to the traitlet in the other plugin makes sense, then I guess at least a comment in both places mentioning the logic-duplication would help prevent mistakes in the future.

pllim replied:
Do you want me to set this traitlet from the Slice plugin instead? I think I can do that.

kecnry replied:
No, then that would require the slice plugin checking to see if this plugin exists and pushing the value, which for some reason feels more against the whole idea of modular plugins than subscribing to changes 🤷‍♂️.  Whatever you think is best is fine, I won't insist.

pllim replied:
Let's leave it as-is for now then, since it works. We can always have a follow-up PR to make it cleaner if desired. Thanks. I appreciate your feedback nonetheless.

pllim replied:
p.s. I will add comment to Slice as you requested.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +112,184 @@ def vue_save_figure(self, filetype):
+                video = cv2.VideoWriter(filename, cv2.VideoWriter_fourcc(*'mp4v'), fps, frame_size, True)  # noqa: E501
+                for cur_pngfile in temp_png_files:
+                    video.write(cv2.imread(cur_pngfile))
+        finally:
```
do we at least need a catch-all except which resets `self.movie_recording = False`?  If I try to run with the spectrum viewer, it raises a snackbar but all buttons are disabled and killing doesn't do anything since there isn't an active thread.

pllim replied:
Maybe. I keep forgetting about the spectrum viewer. Even better if we can disable the button when spectrum viewer is selected but I don't know how to do that.

kecnry replied:
Is there a reason we can't just support the spectrum viewer?

But either way, the try-finally currently swallows all errors (not just caused by selecting the spectrum viewer) and results in a broken state.

pllim replied:
Spectrum viewer does not have bqplot save PNG feature because of this:

https://github.com/spacetelescope/jdaviz/blob/a9f13e9cad6de60e5a96611651bf8207e8dd76a7/jdaviz/configs/specviz/plugins/viewers.py#L9

It comes from glue-jupyter, not bqplot. To support it, I think we need to implement stuff over at glue-jupyter. Also, is it really useful to have a movie that just showing the slider slowly... sliding?

pllim replied:
I added a check to disable the "Export to MP4" button if spectrum-viewer is selected.

> If I try to run with the spectrum viewer, it raises a snackbar but all buttons are disabled and killing doesn't do anything since there isn't an active thread.

It should not disable the buttons at all when you try to record with spectrum viewer, as the exception would be raised before the "movie_recording" flag is even set in the thread because like you say, no thread would be generated, so I have no idea how you got to that state in the first place. But no matter, with the check above, this should not be possible anymore from the GUI.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +112,184 @@ def vue_save_figure(self, filetype):
+                if os.path.exists(cur_pngfile):
+                    os.remove(cur_pngfile)
+
+        if self.movie_kill:
+            if os.path.exists(filename):
+                os.remove(filename)
+            self.movie_kill = False
```
what would happen if we didn't have this?  Would the file be corrupted or just end at whatever frame it was interrupted?

pllim replied:
Not sure. Depends on when the thread is killed. So better to just clean it up.

kecnry replied:
Ok, then I think that should be included in the tooltip for the stop button.  Otherwise it might be reasonable to expect that you can use it to just decide the end point while it's running.

pllim replied:
OK I added a tooltip. I copied the code from Slice plugin and the tooltip seems to only show when the button is not disabled. Good enough?

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +112,184 @@ def vue_save_figure(self, filetype):
+
+        # No wrapping. Forward only.
+        slice_plg = self.app._jdaviz_helper.plugins["Slice"]._obj
+        if i_start < 0:  # pragma: no cover
+            i_start = 0
+        if i_end > slice_plg.max_value:  # pragma: no cover
+            i_end = slice_plg.max_value
```
should these cases be caught in UI validation or just done silently?  If there is no validation, the user might expect negative indices to work (or maybe we could support negative indices?).

pllim replied:
I don't see why someone would want to record a movie backwards (unless they making a horror movie). Anyways, I added validation. I wouldn't worry about negative index support unless we have real users asking for it.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.py:
```diff
@@ -74,3 +112,184 @@ def vue_save_figure(self, filetype):
+                f"inclusive, at {self.movie_fps} FPS.",
+                sender=self, color="success"))
+
+    def vue_kill_recording(self, *args):  # pragma: no cover
+        self.movie_kill = True
+        self.hub.broadcast(SnackbarMessage(
+            f"Movie recording interrupted by user, {self.movie_filename} will be deleted.",
+            sender=self, color="warning"))
```
It might be nice to have this exposed in the public API as well (or to have `save_movie` be blocking and a keyboard interrupt to trigger this).  Although if public, `interrupt` or `stop` might be a bit... friendlier... than "kill".

pllim replied:
I don't feel like making it too easy to kill with API. That button is supposed to be a last resort button.

pllim replied:
But I will rename.

kecnry commented on jdaviz/configs/default/plugins/export_plot/export_plot.vue:
```diff
@@ -25,11 +26,86 @@
+                  ></v-text-field>
+                </v-col>
+              </v-row>
+              <v-row v-if="movie_msg===''" justify='end'>
+                <v-btn
+                 color="primary"
+                 @click="() => save_movie('mp4')"
+                 :disabled="movie_recording"
+                >
+                 Export to MP4
+                </v-btn>
+                <v-btn color="primary" icon @click="kill_recording" :disabled="!movie_recording">
+                  <v-icon>stop</v-icon>
+                </v-btn>
+              </v-row>
```
both of these would benefit from tooltips.  Should the stop button be disabled or not visible at all before pressing the export button?

pllim replied:
The stop button is disabled when movie isn't recording (gray/unclickable vs blue/clickable). I guess I'll just hide it too as you suggested.

duytnguyendtn commented on jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:
```diff
@@ -137,20 +137,29 @@ def _write_moment_to_fits(self, overwrite=False, *args):
 
-        path = Path(self.filename).resolve()
-        if path.exists():
+        # Make sure file does not end up in weird places in standalone mode.
+        path = os.path.dirname(self.filename)
+        if path and not os.path.exists(path):
+            raise ValueError(f"Invalid path={path}")
+        elif (not path or path.startswith("..")) and os.environ.get("JDAVIZ_START_DIR", ""):  # noqa: E501 # pragma: no cover
+            filename = Path(os.environ["JDAVIZ_START_DIR"]) / self.filename
+        else:
+            filename = Path(self.filename).resolve()
+
+        if filename.exists():
```
I'd REALLY like to architect a pathlib-only solution here and avoid using `os.path`. I think we could make this codeblock much more elegant with a combination of `path.exists()`, `path.parent` and `path.is_dir()`, `path.relative_to()`, and `path.is_absolute()`

I'm generally a stickler for path handling and I admit I was tempted to leave a "changes requested" review here, but I'll bite my tongue and allow myself to get overruled here since we're at the end of the sprint

duytnguyendtn reviewed (commented):
Confirmed moment map writes to the starting directory on standalone! Thanks for the fix. Please see my comment

pllim commented on jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:
```diff
@@ -137,20 +137,29 @@ def _write_moment_to_fits(self, overwrite=False, *args):
 
-        path = Path(self.filename).resolve()
-        if path.exists():
+        # Make sure file does not end up in weird places in standalone mode.
+        path = os.path.dirname(self.filename)
+        if path and not os.path.exists(path):
+            raise ValueError(f"Invalid path={path}")
+        elif (not path or path.startswith("..")) and os.environ.get("JDAVIZ_START_DIR", ""):  # noqa: E501 # pragma: no cover
+            filename = Path(os.environ["JDAVIZ_START_DIR"]) / self.filename
+        else:
+            filename = Path(self.filename).resolve()
```
Yeah, I have the same sentiment but refactoring moment map is out of scope here. Moment map is already like that when I joined the project, so not sure what was the original intention of the non-pathlib design. Keep in mind also that `astropy.io.fits` predates pathlib, so sometimes using non-pathlib is unavoidable.

pllim replied:
Unless you mean specifically that you do not like `path = os.path.dirname(self.filename)`? That is just to check the string content of the path though. It is not used in the actual parsing, as you can see.

kecnry reviewed (approved):
Looks good to me now, thanks!  I'm guessing with the JS-callback issues in test cases, it would be difficult to increase the test coverage significantly?  Docs will also need to be fixed before merge, but that seems like it will only be a minor change, so I'll approve now.

When doing this final review, I also found a few more styling inconsistencies, but since that affects multiple plugins, I'll just open a follow-up PR after this gets merged.

rosteen commented on jdaviz/configs/default/plugins/export_plot/export_plot.vue:
```diff
@@ -25,11 +26,110 @@
        </v-list-item>
       </v-list>
+
+      <v-row v-if="config==='cubeviz'">
```
```suggestion
     <v-row v-if="config==='cubeviz' && viewer_selected!='spectrum-viewer'">
```

This would hide the `Export to Video` expansion tray completely if `spectrum-viewer` is selected - I kind of like that better than just graying-out the export button in that case, I could see it being a little confusing for the user that all the options are there but they can't actually hit the button. I don't feel strongly about this though and would be ok merging with the current behavior.

rosteen reviewed (approved):
I left one suggestion, but consider that optional so I'm approving. Nice work!
